[test ui] Ui stage locking

Bjvanminnen opened the pull request:
This is a pretty sizable set of changes for the spec here: https://docs.google.com/document/d/1YLaC0FpYJrpnBtWSs1wr27brTbrBoKxUw-rC46FoGuI/edit

The general idea is that levelbuilders can mark stages as lockable. It's only expected to be used for assessment stages for the time being. Lockable stages are hidden, unless you're an approved teacher, or the student of an approved teacher. 

Lockable stages default to being locked. Teachers can then unlock them on a per-student basis. Once they are unlocked, they will remain so for 24 hours, at which point they will become locked again. Teachers can also put stages into a readonly mode in which students are able to view the answers. Again, this will only last for 24 hours, at which point the stage will again become locked.

This PR is currently lacking in test coverage. This is in part because we're in a rush to get this feature in. My plan is to start writing tests, but not necessarily block committing this on them being finished.

![image](https://cloud.githubusercontent.com/assets/1767466/17413262/6634c744-5a35-11e6-9598-4648380d5463.png)

![image](https://cloud.githubusercontent.com/assets/1767466/17413266/6b32af0e-5a35-11e6-9556-56f19025ba19.png)

![image](https://cloud.githubusercontent.com/assets/1767466/17413282/73818b58-5a35-11e6-8826-14eae0e6ec77.png)

![image](https://cloud.githubusercontent.com/assets/1767466/17413982/57514d12-5a38-11e6-8ae3-92e406933801.png)

![Uploading image.png…]()

Bjvanminnen commented on dashboard/app/controllers/api_controller.rb:
```diff
@@ -11,6 +11,51 @@ def user_hero
+
+    render json: data
+  end
+
```
Is api_controller the right place for this logic to live? If not, where would it make more sense?

joshlory replied:
Recommend the main logic be pushed down to either the `Stage` or `Section` model, whichever is appropriate.  Makes it easier to test each unit of functionality.

Bjvanminnen commented on dashboard/app/views/levels/_locked.html.haml:
```diff
@@ -0,0 +1,3 @@
+.locked-level{style: 'font-size: 14px; margin-top: 10px; margin-bottom: 10px;'}
+  =t('locked_stage')
+= link_to t('continue'), @script_level.next_level_or_redirect_path_for_user(@current_user), class: 'btn btn-large btn-primary next-stage submitButton'
```
Is it silly to create a partial for something this small that is used in only one place?

mehalshah replied:
I suggest removing this

Bjvanminnen commented on dashboard/config/routes.rb:
```diff
@@ -340,6 +340,8 @@ module OPS
   get '/dashboardapi/v1/pd/k5workshops', to: 'api/v1/pd/workshops#k5_public_map_index'
 
+  post '/api/lock_status', to: 'api#update_lockable_state'
+  get '/api/lock_status', to: 'api#lockable_state'
```
Are these reasonable routes?

mehalshah replied:
Yes

Bjvanminnen commented on dashboard/db/migrate/20160802170542_add_readonly_unlocked_at_to_user_levels.rb:
```diff
@@ -0,0 +1,6 @@
+    add_column :user_levels, :view_answers, :boolean
+    add_column :user_levels, :unlocked_at, :datetime
+  end
+end
```
This is a big migration that we probably want to run separately from this rest of this PR once ready. @breville @ashercodeorg had some thoughts/experience on how to best manage this.

breville replied:
Yes, I'd recommend breaking it out into a separate PR so that you can run it in isolation on production after hours, when we disable writes to UserLevel using the queueing mechanism, and draining it after the migration completes.  We went through something similar with https://github.com/code-dot-org/code-dot-org/pull/6973/files

ashercodeorg replied:
Can these columns be nil? Is there a default value for `view_answers`?

Bjvanminnen replied:
They can be nil. In my local db it looks like view_answers defaults to NULL everywhere, which gets treated as false afaict, which is the desired behavior. Would it be better to just set to 0 (false) everywhere? If so, what's the syntax to do that as part of the migration?

islemaster commented on apps/src/code-studio/components/progress/ScriptTeacherPanel.js:
```diff
@@ -0,0 +1,148 @@
+    } = this.props;
+    const hasSections = Object.keys(sections).length > 0;
+    // TODO - i18n
+    // TODO - don't forget section is conditional on having unlocked stages
```
TODOs?

Bjvanminnen replied:
Thanks. Thought I'd found/fixed all these but apparently missed this.

islemaster commented on apps/src/code-studio/components/progress/ScriptTeacherPanel.js:
```diff
@@ -0,0 +1,148 @@
+                <button value={ViewType.Teacher}>Teacher</button>
+              </ToggleGroup>
+            </div>
+          </div>
```
This label/control could be easily extracted to a `ViewAsToggle` component.  Even inline in this file, it'd probably improve the readability of the render method.

islemaster commented on apps/src/code-studio/components/progress/StageLock.jsx:
```diff
@@ -0,0 +1,87 @@
+const StageLock = React.createClass({
+  propTypes: {
+    sectionsLoaded: React.PropTypes.bool.isRequired,
+    unlocked: React.PropTypes.bool.isRequired, // TODO - connect?
```
TODO?

islemaster commented on apps/src/code-studio/components/progress/ScriptTeacherPanel.js:
```diff
@@ -0,0 +1,148 @@
+              </ToggleGroup>
+            </div>
+          </div>
+          {!sectionsLoaded && <div style={styles.text}>Loading...</div>}
```
Looks like already set up `commonMsg.loading()` and are using it in StageLock.jsx; use here too?

islemaster commented on apps/src/code-studio/components/progress/StageLockDialog.jsx:
```diff
@@ -0,0 +1,284 @@
+import React from 'react';
+import Radium from 'radium';
+import { connect } from 'react-redux';
+import BaseDialog from '@cdo/apps/templates/BaseDialog';
```
Is there a particular reason you're using `BaseDialog` and not the more opinionated `Dialog` @pcardune recently introduced?

Bjvanminnen replied:
I don't need any of the elements of Dialog (which wraps BaseDialog). For example, here is my dialog
![image](https://cloud.githubusercontent.com/assets/1767466/17491358/2d98d59a-5d5c-11e6-99fa-b18c57f023e3.png)
and here is a `Dialog`
![image](https://cloud.githubusercontent.com/assets/1767466/17491379/3dc91628-5d5c-11e6-8fa0-b1d92bd7403a.png)

I don't have a title or icon, and my save/cancel buttons are different. Using BaseDialog got me the parts that I needed (hide/close, x in the corner, and a backdrop).

islemaster replied:
Just as a design choice, it seems like you _could_ have some elements in common.  The text of your save/cancel buttons is different, but is there a reason we're not using the `Dialog` button styles?  Part of the motivation behind `Dialog` was to standardize our UI as much as possible.

It does seem like we should have a 'teacher styles' version - the default `Dialog` styles are pretty large and student-friendly, and don't necessarily lend themselves to an information-dense view.  Still, I think things like colors and button position could be more standard - it's odd that you've used call-to-action orange on the assessment steps buttons, but not on the save button, and the gray 'Cancel' button looks like it could be disabled.

It's probable that a more design-y person should be involved in this discussion.  In any case, don't let any of this block the PR.

Bjvanminnen replied:
tagging @poorvasingal and @tanyaparker if they want to weigh in on the design decisions.

Note - one other thing that's pretty different about my save button (in addition to the different color), is that it is an async operation and changes to "Saving..." while performing the operation. I guess we could make `Dialog` know how to handle this sort of thing too if we wanted.

islemaster replied:
Generally speaking, it just takes an `onConfirm` callback that can do what
you want.  It also supports swapping in your own button components if you
prefer that (example from storybook here:
https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/templates/Dialog.jsx#L227
).

On Mon, Aug 8, 2016 at 11:52 AM, Brent Van Minnen notifications@github.com
wrote:

> In apps/src/code-studio/components/progress/StageLockDialog.jsx
> https://github.com/code-dot-org/code-dot-org/pull/9760#discussion_r73932383
> :
> 
> > @@ -0,0 +1,284 @@
> > +import React from 'react';
> > +import Radium from 'radium';
> > +import { connect } from 'react-redux';
> > +import BaseDialog from '@cdo/apps/templates/BaseDialog';
> 
> tagging @poorvasingal https://github.com/poorvasingal and @tanyaparker
> https://github.com/tanyaparker if they want to weigh in on the design
> decisions.
> 
> Note - one other thing that's pretty different about my save button (in
> addition to the different color), is that it is an async operation and
> changes to "Saving..." while performing the operation. I guess we could
> make Dialog know how to handle this sort of thing too if we wanted.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> https://github.com/code-dot-org/code-dot-org/pull/9760/files/b74016256c91abfec19d89261415d5b56b0c8c7f#r73932383,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/ABinkeFu9OP-tdR9ThjaX92LqjLu2Sx4ks5qd3r5gaJpZM4JZxpO
> .

## 

Brad Buchanan
Software Engineer, Code.org
brad@code.org

tanyaparker replied:
We don't usually see orange buttons for teachers on the teacher dashboard. I can't actually think of a case. The moving students dialog (and saving data for sections) set the precedent for blue Save button. And in this dialog the orange buttons for locking/unlocking are actually the buttons we want them to use more, rather than saving the exceptions. Although I agree it's odd to have 4 orange buttons. But would also be weird if we had only 1 out of the 4 to be orange and have it change on some kind of rotation as we track which one you clicked last and encourage you to click the next one in the sequence. I think this is best choice given our current constraints, and when we start our designer (really really soon) have them take a crack at this.

islemaster replied:
Ah - see, I also didn't realize that "Save" didn't apply to the options in the top half of the dialog.

tanyaparker replied:
That's good feedback too. This dialog is a little awkward because they're somewhat related but not directly. Like you wouldn't need to hit "Lock" AND "Save"

poorvasingal replied:
@tanyaparker - might be misunderstanding your comment but you do need to hit both.  "Lock stage" just puts everything in the "locked" column.  You'd still need to hit "save" for your changes to apply.

tanyaparker replied:
Oh right. I got confused myself. I forgot that clicking the button will give you the immediate feedback of changing the table, but then you still need to click Save to save the changes of the table. Perhaps in the instructions we should say "don't forget to save your changes each time" or something like that. Because I just had a momentary blip thinking that clicking the action button was all I had to do.

poorvasingal replied:
Playing with the feature on adhoc...I actually don't think we need those extra instructions.  The dialog stays up until you hit "save" or "cancel" and you see the table cells changing.  So it becomes obvious that you need to hit Save.

tanyaparker replied:
Then we should get rid of the x to close and make sure you can't exit by clicking outside of the dialog.

islemaster replied:
I added an optional `uncloseable` property to `BaseDialog` a while back that does exactly this.

islemaster commented on apps/src/code-studio/components/progress/StageLockDialog.jsx:
```diff
@@ -0,0 +1,284 @@
+
+  getInitialState() {
+    return {
+      lockStatus: this.props.initialLockStatus
```
[Props in getInitialState is an antipattern](https://facebook.github.io/react/tips/props-in-getInitialState-as-anti-pattern.html).  It looks like you're using state here to manage temporary local state, separate from the save process?  Why can't/shouldn't redux handle all of this?

islemaster replied:
My bad here, docs say "it's not an anti-pattern if you make it clear that the prop is only seed data for the component's internally-controlled state:"  I'm still a little curious about the separation though.

Bjvanminnen replied:
"However, it's not an anti-pattern if you make it clear that the prop is only seed data for the component's internally-controlled state:"
My goal in naming this initialLockStatus was to meet the above criteria.
It's possible this should all be in redux, but that means creating a bunch more actions for stuff that only this component cares about. I'll think a little on whether that's worth it or not.

Bjvanminnen replied:
Part of the reason for the separation is that this component is only changing local state. The lock status itself doesn't change until/unless we hit save, so it felt appropriate to have this state live inside the component.

islemaster commented on apps/src/code-studio/components/progress/StageLockDialog.jsx:
```diff
@@ -0,0 +1,284 @@
+                <th style={styles.headerRow}>{commonMsg.locked()}</th>
+                <th style={styles.headerRow}>{commonMsg.editable()}</th>
+                <th style={styles.headerRow}>{commonMsg.answersVisible()}</th>
+              </tr>
```
Why is your header row inside `tbody`?  More generally, have you seen [`Reactabular`](http://reactabular.js.org/)?

Bjvanminnen replied:
TODO (brent): revisit this comment

islemaster commented on apps/src/code-studio/components/progress/course_progress_row.jsx:
```diff
@@ -79,11 +80,13 @@ const CourseProgressRow = React.createClass({
 
   render() {
-    const stage = this.props.stage;
+    const { stage, unlockedStageIds} = this.props;
```
Funky spacing here.

islemaster commented on apps/src/code-studio/progress.js:
```diff
@@ -93,6 +103,35 @@ progress.renderCourseProgress = function (scriptData, currentLevelId) {
+ * Query the server for progress of all students in section, and render this to
+ * our teacher panel.
+ */
+// TODO - rename since this is more than teacher panel (it's lockable stuff everywhere)
```
TODO rename.

Bjvanminnen replied:
Hah. I could have sworn I had addressed some of these TODOs. Turns out I had, but hadn't pushed my commit :(

islemaster commented on apps/src/code-studio/stageLockRedux.js:
```diff
@@ -0,0 +1,240 @@
+      sections: action.sections,
+      sectionsLoaded: true,
+      selectedSection: sectionId,
+      unlockedStageIds: unlockedStages(currentSection)
```
It looks like this information can be derived from other information on the state.  Is there a reason we're storing this instead of calculating it as-needed?

islemaster commented on apps/src/code-studio/stageLockRedux.js:
```diff
@@ -0,0 +1,240 @@
+    }));
+
+    if (saveData.length === 0) {
+      dispatch(closeLockDialog());
```
I was surprised to find that this action also closes the dialog.  It might be cleaner for this action creator just to take an `onComplete` callback, and let the caller specify that the dialog should be closed when the save is done.

islemaster commented on apps/src/code-studio/stageLockRedux.js:
```diff
@@ -0,0 +1,240 @@
+      dispatch(closeLockDialog());
+    });
+  };
+};
```
Nice use of `redux-thunk` here :+1:

islemaster commented:
I don't see any tests!

islemaster commented:
There's a mix of styles-from-CSS and inline styles in the React components.  Is this because you're integrating with UI that's mostly driven by dashboard?  Not sure it's worth cleaning up yet, but want to get your thoughts.

islemaster commented:
Would be nice to add a StageLockDialog entry to react-storybook.

Bjvanminnen commented:
See PR description for comment about tests (essentially they're the first thing on my list to do). I like the idea of also adding something to react-storybook, though that's not something I'd like to necessarily block this PR on.

Regarding mix of classNames and inline styles, it's because teacher-panel is something that exists elsewhere already, and is defined in haml. I didn't want to do another haml view here, and so I did it in react, but use some of the same classes. It's possible I could/should still use inlined versions of those here.

islemaster commented:
JS LGTM

joshlory commented on apps/src/code-studio/activityUtils.js:
```diff
@@ -2,34 +2,53 @@
 const MINIMUM_PASS_RESULT = 20;
 const MINIMUM_OPTIMAL_RESULT = 30;
 export const SUBMITTED_RESULT = 1000;
+export const LOCKED_RESULT = 1001;
```
The ordering here has some implications: a higher value takes precedence over a lower value, and values below 1000 should't be cached by the client.  Is this the correct ordering?  We should also leave room for new values if you think there could ever be a status with higher precedence than `submitted` but lower precedence than `locked`.

It's an unfortunate system, but these values are written to the `Activity` and `UserLevel` table which we can't feasibly migrate.

Bjvanminnen replied:
In this case, we should never submit with LOCKED_RESULT (such that this shouldn't end up in Activity at all I don't think). It also shouldn't end up in UserLevel table. I think this is true for SUBMITTED_RESULT as well. Instead, both of these are used to track some client side state.

I think we're kind of overloading what result means in both cases, which isn't super desirable. This is done in part because this is the only way we currently have to track user-specific data. I would love to clean some of this up in the future, but I don't think adding LOCKED_RESULT (which is fairly similar to how SUBMITTED_RESULT works) is making things significantly worse.

joshlory commented on apps/src/code-studio/components/progress/ScriptTeacherPanel.js:
```diff
@@ -0,0 +1,164 @@
+  }
+};
+
+const ViewAsToggle = ({viewAs, setViewType}) => (
```
Is this a React class?  I haven't seen this syntax before.

Bjvanminnen replied:
Yep, this is a functional stateless component https://facebook.github.io/react/docs/reusable-components.html#stateless-functions

joshlory replied:
Cool, TIL!

joshlory commented on apps/src/code-studio/components/progress/course_progress.jsx:
```diff
@@ -24,7 +24,7 @@ const CourseProgress = React.createClass({
     professionalLearningCourse: React.PropTypes.bool,
     focusAreaPositions: React.PropTypes.arrayOf(React.PropTypes.number),
-    stages: React.PropTypes.arrayOf(stageShape),
+    stages: React.PropTypes.arrayOf(stageShape)
```
Note that this comma was previously added in https://github.com/code-dot-org/code-dot-org/pull/9293/files#diff-a04b10d3f83ac4aa14e208f43bdabeb7R27.

I don't have any preference for trailing commas or not, but churning back and forth seems bad 😄.

Bjvanminnen replied:
I've actually recently developed a slight preference for trailing commas. Not sure why I chose to remove these.

joshlory commented on apps/src/code-studio/components/progress/progress_dot.jsx:
```diff
@@ -185,7 +187,8 @@ export const ProgressDot = React.createClass({
     const isPeerReview = level.kind === 'peer_review';
     const iconForLevelStatus = !isUnplugged && this.props.courseOverviewPage && this.getIconForLevelStatus(level);
-    const levelUrl = level.locked ? undefined : level.url + location.search;
+    const isLocked = level.locked || level.status === LevelStatus.locked;
```
Why have both `level.locked` and `level.status === LevelStatus.locked`?

Bjvanminnen replied:
Could have sworn I responded to this comment already, but perhaps it didn't go through because of the internet.

`level.locked` is a per level user-agnostic concept, I believe only used by peer reviews ( @mehalshah can probably confirm), which I believe we find out about as part of page load in `scriptData` . `level.status === LevelStatus.locked` is a user-specific concept, which we don't know about until we request user progress.

Not sure if there's a better way of disambiguating these two concepts.

mehalshah commented on dashboard/app/helpers/levels_helper.rb:
```diff
@@ -594,4 +594,9 @@ def can_view_solution?
   end
+
+  def should_show_inline_answer?
+    current_user.try(:authorized_teacher?) && !@script.try(:pd?) ||
```
script.try(:professional_course?)

joshlory replied:
Why even `.try`?

joshlory replied:
I see, `@script` can be nil?

Bjvanminnen replied:
Not certain of what you mean by this comment. Is it that I should be checking `professional_course?` instead of just `pd?`?

joshlory commented on dashboard/app/controllers/api_controller.rb:
```diff
@@ -11,6 +11,87 @@ def user_hero
+      user_level_data = item[:user_level_data]
+
+      if user_level_data[:user_id].nil? || user_level_data[:level_id].nil? || user_level_data[:script_id].nil?
+        raise 'Must provide user, level, and script ids'
```
Does this turn into a 5xx error?  `400 Bad Request` may be more appropriate.

Bjvanminnen replied:
I had assumed it would, but testing now it looks like we get back a 404, and the responseText contains the call stack (though I suspect that's only because this is a dev environment). Is there a better way of doing this such that we get back 500s?

joshlory replied:
5xx errors will be logged to HoneyBadger.  If we think we'll catch bugs this way it may be appropriate to log.  The risk is if it contributes to the noise in HoneyBadger without actionable repros.

Bjvanminnen replied:
I think 500 errors are probably desirable here. What's the syntax to get it to respond with a 500?

joshlory replied:
From an API perspective, 4xx (client error) seems more appropriate than 5xx (server error).  The server is capable of handling the request, but couldn't because the client gave bad input.

joshlory commented on dashboard/app/controllers/api_controller.rb:
```diff
@@ -11,6 +11,87 @@ def user_hero
+      unless User.find(user_level_data[:user_id]).teachers.include? current_user
+        raise 'Can only update lockable state for users students'
+      end
+
```
Above here is validation that makes sense in the controller.  Can the logic below here be moved to a method on the `UserLevel` model (and tested in a rake test)?

joshlory commented on dashboard/db/migrate/20160809170542_add_view_answers_unlocked_at_to_user_levels.rb:
```diff
@@ -0,0 +1,6 @@
+class AddViewAnswersUnlockedAtToUserLevels < ActiveRecord::Migration
+  def change
+    add_column :user_levels, :view_answers, :boolean
+    add_column :user_levels, :unlocked_at, :datetime
```
A heads up, last time we migrated the `UserLevel` table we had to pause writes (even then it took a few hours).  +cc @breville who ran the last migration to add `:submitted` and @ashercodeorg.

Are there other changes we should consider making at the same time?  E.g. we've talked about whether the `:submitted` boolean should maybe be an enum.

ashercodeorg replied:
Another possible change might be adding a `total_time` column. Today, we track (client-side) elapsed time on a level, saving it to `activities.time`. If we deprecate the `activities` table, I imagine there being value in having `user_levels.total_time`. What do you think @joshlory @ryansloan @Hamms?

joshlory commented on dashboard/app/helpers/levels_helper.rb:
```diff
@@ -594,4 +594,9 @@ def can_view_solution?
     end
   end
+
+  def should_show_inline_answer?
```
Is this a property of the `Level`?  `ScriptLevel`?  `User`?  Prefer it on the model, e.g.:

``` ruby
current_user.can_see_inline_answer? @script_level
# or
@script_level.should_show_inline_answer_to current_user
```

joshlory commented on dashboard/app/helpers/users_helper.rb:
```diff
@@ -91,11 +91,20 @@ def summarize_user_progress_for_all_scripts(user)
           completion_status = activity_css_class(ul)
           submitted = !!ul.try(:submitted)
-          if completion_status != 'not_tried'
+          view_answers = !!ul.try(:view_answers)
```
Don't need to `.try` here if `:view_answers` is a column on `UserLevel`.  If you're concerned about the nil case you can make the column non-nullable.

Bjvanminnen replied:
I think we need to try because ul might be nil.

joshlory replied:
Ah you're right, sorry.  Maybe cleaner to wrap this whole section in `if ul`.

joshlory commented on dashboard/app/views/layouts/application.html.haml:
```diff
@@ -10,6 +10,7 @@
       %meta{'http-equiv' => 'X-UA-Compatible', content: 'IE=Edge'}
     = stylesheet_link_tag    'application', media: 'all'
     = javascript_include_tag 'application'
+    = javascript_include_tag "js/#{js_locale}/common_locale"
```
Why this change?  Will this cause `common_locale` to be double-included on puzzle pages?

Bjvanminnen replied:
The change was because we have apps strings we need in code-studio, i.e. wherever code-studio-common is included. This was resulting in a double inclusion, which I fixed by removing from _apps_dependencies, and then adding to embed.

joshlory commented on dashboard/app/views/levels/_level_group.haml:
```diff
@@ -27,50 +27,53 @@
-    - unless @script_level.nil?
-      = render partial: 'levels/dialog', locals: {app: app, data: data, previous_button: page.page_number > 1, next_button: page.page_number < @total_page_count}
+  - if @current_user.user_level_locked?(@script_level, @level)
+    = render partial: 'levels/locked'
```
Are level groups the only level type that can be locked?

joshlory commented on dashboard/db/migrate/20160809170542_add_view_answers_unlocked_at_to_user_levels.rb:
```diff
@@ -0,0 +1,6 @@
+class AddViewAnswersUnlockedAtToUserLevels < ActiveRecord::Migration
+  def change
+    add_column :user_levels, :view_answers, :boolean
```
Should this boolean column be non-nullable?

Bjvanminnen replied:
I currently treat false and NULL to be equivalent. Would it be better to say it's non-nullable and have a default value of false?

Bjvanminnen replied:
I currently treat false and NULL to be equivalent. Would it be better to say it's non-nullable and have a default value of false?

joshlory replied:
Yep, better to avoid the three-state boolean problem when we really want there to be only two states: https://robots.thoughtbot.com/avoid-the-threestate-boolean-problem

Bjvanminnen commented:
I've addressed a bunch of feedback, and also fixed a couple of bugs that we found. Would be great if I could get another look at this @islemaster @joshlory, though I'm more interested in feedback on the dashboard side of things than the apps stuff. Hoping to merge this later today unless we find any big issues.

islemaster commented:
Didn't notice anything in a quick scan, except you could use the `uncloseable` property on `BaseDialog` if you want to force people to use the save/cancel buttons.

ashercodeorg commented on dashboard/app/controllers/api_controller.rb:
```diff
@@ -11,6 +11,54 @@ def user_hero
+      end
+
+      unless User.find(user_level_data[:user_id]).teachers.include? current_user
+        # Can only update lockable state for users students
```
nit: s/users/user's

ashercodeorg commented on dashboard/app/models/user_level.rb:
```diff
@@ -82,8 +92,35 @@ def paired?
+
+  def has_autolocked?(stage)
+    return false unless stage.lockable?
+    (self.unlocked_at && self.unlocked_at < AUTOLOCK_PERIOD.ago)
```
nit: Remove the superfluous parens.

joshlory commented on dashboard/app/controllers/activities_controller.rb:
```diff
@@ -56,6 +56,14 @@ def milestone
+      if user_level.nil? || user_level.locked?(@script_level.stage)
+        return head 403
+      end
+    end
```
I wonder if this should be in `User.track_level_progress_sync`?  As it's written now, anything that stores progress without posting to `/milestone` will bypass the `user_level.locked?` check.

Bjvanminnen replied:
If done in `track_level_progress_sync` I think we can get into trouble if the message ends up being queued. i.e. I submit a milestone request while unlocked, the request enters the queue, the level gets locked, my message gets dequeued, and now it fails because the level is locked (even though it wasn't when I submitted).

I could move to `track_progress_for_user`, but it looks like that is only called by the milestone anyways?

joshlory replied:
I think we expect `LevelGroup` to be only used in non-cacheable scripts, but I don't know that this is enforced in code.

joshlory commented on dashboard/app/controllers/api_controller.rb:
```diff
@@ -11,6 +11,54 @@ def user_hero
+
+  # For a given user, gets the lockable state for each student in each of their sections
+  def lockable_state
+    return unless current_user
```
Should this be:

``` ruby
unless current_user
  render json: {}
  return
end
```

joshlory commented on dashboard/app/controllers/api_controller.rb:
```diff
@@ -11,6 +11,54 @@ def user_hero
+
+    data = current_user.sections.each_with_object({}) do |section, section_hash|
+      next if section[:deleted_at]
+      @section = section
```
When there are multiple sections, `@section` will be set to the last section?

joshlory replied:
Not used?

Bjvanminnen replied:
Used by load_script

joshlory commented on dashboard/app/helpers/application_helper.rb:
```diff
@@ -65,6 +65,8 @@ def best_activity_css_class(user_levels)
     elsif result == Activity::REVIEW_ACCEPTED_RESULT
       'review_accepted'
+    elsif user_level.try(:locked)
+      'locked'
```
Confirming that this is the correct precedence in this list — if a peer review level is ever lockable for some reason, the peer review status will take precedence over the locked status.

Bjvanminnen replied:
The precedence shouldn't matter as these two features should be mutually exclusive.

joshlory commented on dashboard/app/models/stage.rb:
```diff
@@ -124,4 +124,30 @@ def summarize
+  def lockable_state(students)
+    return unless self.lockable?
+
+    # assumption that lockable selfs have a single (assessment) level
```
All of this seems pretty specific to `LevelGroup`.  Is the plan to expand this in the future to apply to stages with other level types in them?

Otherwise a lot of this logic might make more sense on the `LevelGroup` model (probably too late for this feedback).

Bjvanminnen replied:
I think when we started this feature there was a bit of a hand-wavy not fully thought out desire that this might be something we wanted to support more broadly, but over time we focused on the feature we knew we needed to support, which is LevelGroup.

If it's okay to have it here for now, I'm okay with taking on a task to see if it might make sense to move some of this logic into LevelGroup in a separate PR.

Bjvanminnen replied:
https://codeorg.axosoft.com/viewitem?id=156&type=features&force_use_number=true

joshlory commented on dashboard/app/models/user_level.rb:
```diff
@@ -82,8 +92,35 @@ def paired?
 
   def handle_unsubmit
-    if submitted_changed? from: true, to: false
+    if submitted_changed? from: true, to: false && !self.readonly_answers
```
Why don't change `best_result` when `readonly_answers`?

Bjvanminnen replied:
This case should actually never happen, as submitted being/becoming false and readonly_answers being true is impossible (which is now validated/enforced). I think I probably had a case where I was hitting this condition before this was enforced. I'll remove the `&& !self.readonly_answers` bit from this, and make sure nothing breaks.

joshlory commented on dashboard/app/views/levels/_match_answer.html.haml:
```diff
@@ -1,5 +1,5 @@
 / Show the correct answer to authorized teachers
-- if current_user.try(:authorized_teacher?) && !@script.try(:pd?)
+- if @current_user.try(:should_see_inline_answer?, @script_level)
```
Why `current_user` -> `@current_user`?

(Is there any risk that a consumer of this view is rendering it without setting `@current_user`?)

Bjvanminnen replied:
No good reason. I'll fix usages.

joshlory commented:
Added some nits, but I don't think any of them are blocking.  LGTM 👍

joshlory commented on dashboard/app/models/user_level.rb:
```diff
@@ -25,19 +25,29 @@
 
+  def readonly_requires_submitted
+    if readonly_answers? && !submitted?
+      errors.add(:readonly_answers, 'readonly_answers only valid on submitted UserLevel')
```
How will this error be surfaced?  My concern is that most places we update `UserLevel` are in async actions, or somewhere the user (or us) won't see the error.

Bjvanminnen commented:
circle-ci failure is an eyes failure. brian says these are ignoreable for the time being, and i looked on applitools and the diffs looked unrelated.
